client/allocrunner/taskrunner/envoy_version_hook.go

```diff
@@ -0,0 +1,170 @@
+	"github.com/pkg/errors"
+)
+
+const (
+	// envoyVersionHookName is the name of this hook and appears in logs.
+	envoyVersionHookName = "envoy_version"
+
+	// envoyLegacyImage is used when the version of Consul is too old to support
+	// the SupportedProxies field in the self API.
+	//
+	// This is the version defaulted by Nomad before v0.13.0 and/or when using versions
+	// of Consul before v1.7.8, v1.8.5, and v1.9.0.
+	envoyLegacyImage = "envoyproxy/envoy:v1.11.2@sha256:a7769160c9c1a55bb8d07a3b71ce5d64f72b1f665f10d81aa1581bc3cf850d09"
+)
+
+type envoyVersionHookConfig struct {
+	alloc         *structs.Allocation
+	proxiesClient consul.SupportedProxiesAPI
+	logger        hclog.Logger
+}
+
+func newEnvoyVersionHookConfig(alloc *structs.Allocation, proxiesClient consul.SupportedProxiesAPI, logger hclog.Logger) *envoyVersionHookConfig {
+	return &envoyVersionHookConfig{
+		alloc:         alloc,
+		logger:        logger,
+		proxiesClient: proxiesClient,
+	}
+}
+
+// envoyVersionHook is used to determine and set the Docker image used for Consul
+// Connect sidecar proxy tasks. It will query Consul for a set of preferred Envoy
+// versions if the task image is unset or references ${NOMAD_envoy_version}. Nomad
+// will fallback the image to the previous default Envoy v1.11.2 if Consul is too old
+// to support the supported proxies API.
+type envoyVersionHook struct {
+	// alloc is the allocation with the envoy task being rewritten.
+	alloc *structs.Allocation
+
+	// proxiesClient is the subset of the Consul API for getting information
+	// from Consul about the versions of Envoy it supports.
+	proxiesClient consul.SupportedProxiesAPI
+
+	// logger is used to log things.
+	logger hclog.Logger
+}
+
+func newEnvoyVersionHook(c *envoyVersionHookConfig) *envoyVersionHook {
+	return &envoyVersionHook{
+		alloc:         c.alloc,
+		proxiesClient: c.proxiesClient,
+		logger:        c.logger.Named(envoyVersionHookName),
+	}
+}
+
+func (envoyVersionHook) Name() string {
+	return envoyVersionHookName
+}
+
+func (h *envoyVersionHook) Prestart(_ context.Context, request *ifs.TaskPrestartRequest, response *ifs.TaskPrestartResponse) error {
+	if h.skip(request) {
+		response.Done = true
+		return nil
+	}
+
+	// We either need to acquire Consul's preferred Envoy version or fallback
+	// to the legacy default. Query Consul and use the (possibly empty) result.
+	proxies, err := h.proxiesClient.Proxies()
+	if err != nil {
+		return errors.Wrap(err, "error retrieving supported Envoy versions from Consul")
+	}
+
+	// Determine the concrete Envoy image identifier by applying version string
+	// substitution (${NOMAD_envoy_version}).
+	image, err := h.tweakImage(h.taskImage(request.Task.Config), proxies)
+	if err != nil {
+		return errors.Wrap(err, "error interpreting desired Envoy version from Consul")
+	}
+
+	// Set the resulting image.
+	h.logger.Trace("setting task envoy image", "image", image)
+	request.Task.Config["image"] = image
+	response.Done = true
```

Comment:
I wonder if we should actually skip this to ensure we re-run it in the event Envoy fails? Seems cheap enough we might as well.

Bootstrap has a similar problem though, so it might not be worth worrying about here if there's a chance we're just going to fail later due to bootstrap changes not being picked up or something.

Reply:
I was thinking the opposite - if Envoy gets restarted the generated envoy config and the envoy version that gets launched should continue to work together. It's kind of interesting in that Envoy will try twice - interpreting the config at versions N and N-1, but it'd be more consistent to just have the version and bootstrap hooks behave the same. OTOH, maybe now it would make sense to query the version and regenerate the config on every sidecar task restart.